### PR TITLE
TClingUtils: Remove HackForDefaultTemplateArg

### PR DESCRIPTION
**Pull request by hahnjo**
This was introduced in 2013 by commit a283dc5847 mentioning `Double32_t`, but all tests pass when removing it. It seems all relevant cases are taken care of by commit c6cfe7cbcd.

### CODE REVIEW

**Comment by Axel-Naumann** [comment]
I suppose that header can be removed, too?

@pcanal what's a test case to look at? GetNormalizedNameWithDefaultTemplateArguments()? ;-)

**Comment by hahnjo** [comment]
> I suppose that header can be removed, too?

Yes, at least two Clang patches can be reverted after this, but I first want to know if it's actually ok to do so before I do that work.

> [...] this case is taken care of by something else

That "something else" could well be commit https://github.com/root-project/root/commit/c6cfe7cbcde6fb4a74abb5b35b2bbd11953f8b7e ...

**Comment by hahnjo** [comment]
> That "something else" could well be commit [c6cfe7c](https://github.com/root-project/root/commit/c6cfe7cbcde6fb4a74abb5b35b2bbd11953f8b7e) ...

Yep, reverting that one finally causes `roottest-root-io-double32-make` to fail!

**Review by Axel-Naumann — APPROVED** [review]

**Comment by vgvassilev** [comment]
In clang we finally have some bits to track the template sugar. If we want to remove this patch we should make sure we resugar the necessary types. I believe we will need to update to clang-16 or backport a bunch of patches.

**Comment by hahnjo** [comment]
> In clang we finally have some bits to track the template sugar. If we want to remove this patch we should make sure we resugar the necessary types. I believe we will need to update to clang-16 or backport a bunch of patches.

We currently have two approaches, I propose to leave https://github.com/root-project/root/commit/c6cfe7cbcde6fb4a74abb5b35b2bbd11953f8b7e in place. Can you give a test that will break by removing `HackForDefaultTemplateArg`?

**Comment by pcanal** [comment]
I am not sure we have good test of the following case:
```
template <typename T> struct myallocator : public std::allocator<T> {};
template <typename T, typename Q = T > class Artificial1 {};
template <typename T, typename Q = std::vector<T> > class Artificial2 {};
template <typename T, typename Q = std::vector<T, myallocator<T> > > class Artificial3 {};
```
where the normalized name for `T==Double32_t` should be:
```
Artificial1<Double32_t, Double32_t>
Artificial2<Double32_t, vector<Double32_t> >
Artificial3<Double32_t, vector<Double32_t, myallocator<Double32_t> > >
```
also:
```
root [7] TClass::GetClass("Artificial3<Double32_t, std::vector<Double32_t, std::allocator<Double32_t> >>")->GetName()
(const char *) "Artificial3<Double32_t,vector<Double32_t> >"
```

i.e. *If* I remember correctly the hack was about adding missing but needed default template parameters.

**Comment by hahnjo** [comment]
Yep, seems to be all working still:
```
root [0] template <typename T> struct myallocator : public std::allocator<T> {};
root [1] template <typename T, typename Q = T > class Artificial1 {};
root [2] template <typename T, typename Q = std::vector<T> > class Artificial2 {};
root [3] template <typename T, typename Q = std::vector<T, myallocator<T> > > class Artificial3 {};
root [4] TClass::GetClass("Artificial1<Double32_t>")->GetName()
(const char *) "Artificial1<Double32_t,Double32_t>"
root [5] TClass::GetClass("Artificial2<Double32_t>")->GetName()
(const char *) "Artificial2<Double32_t,vector<Double32_t> >"
root [6] TClass::GetClass("Artificial3<Double32_t>")->GetName()
(const char *) "Artificial3<Double32_t,vector<Double32_t,myallocator<Double32_t> > >"
root [7] TClass::GetClass("Artificial3<Double32_t, std::vector<Double32_t, std::allocator<Double32_t> >>")->GetName()
(const char *) "Artificial3<Double32_t,vector<Double32_t> >"
```

**Comment by Axel-Naumann** [comment]
Note that this PR (?) seems to cause new errors in the new CI - we were at 0 test failures before, thanks to all your work!

**Comment by hahnjo** [comment]
> Note that this PR (?) seems to cause new errors in the new CI - we were at 0 test failures before, thanks to all your work!

I don't think that's my PR, you can also see it here: https://github.com/root-project/root/pull/13070#issuecomment-1603661635. Looking at the failures, it looks like the CI picks up an old version of `roottest` on Windows. At least the `return 0;` in `test_ROOT9975` was fixed months ago by https://github.com/root-project/roottest/commit/a905ea54cecbe1bb98c2350d06949ba6b8060af1

**Comment by hahnjo** [comment]
*Actually*, the test results say `Windows x86 Debug` which didn't run. So it's uploading a month-old report that it finds somewhere on the filesystem :man_facepalming:

File: interpreter/llvm-project/clang/lib/Sema/SemaTemplate.cpp [hunk 1]
```diff
@@ -39,31 +39,6 @@
...
-bool HackForDefaultTemplateArg::AllowNonCanonicalSubstEnabled = true;
-
-// Default constructor, record the current value
-// of HackForDefaultTemplateArg::AllowNonCanonicalSubst
-// and set it to true.
-HackForDefaultTemplateArg::HackForDefaultTemplateArg() :
-  OldValue(HackForDefaultTemplateArg::AllowNonCanonicalSubstEnabled)
-{
-  AllowNonCanonicalSubstEnabled = true;
-}
```
Comment (hahnjo) [inline interpreter/llvm-project/clang/lib/Sema/SemaTemplate.cpp:51]: Re-reading this hack, doesn't this mean `AllowNonCanonicalSubstEnabled` is always `true`, ie the hack was *always* enabled as opposed to "only with the RAII"?

**Review by vgvassilev — APPROVED** [review]
Seems too good to be true. LGTM!

File: interpreter/llvm-project/clang/lib/AST/ASTContext.cpp [hunk 2]
```diff
@@ -4597,13 +4597,11 @@ QualType ASTContext::getAttributedType(attr::Kind attrKind,
...
-                                        ) const {
-  if (!AllowNonCanonical) {
-     assert(Replacement.isCanonical()
-            && "replacement types must always be canonical");
-  }
+                                         QualType Replacement) const {
+#if 0
+  assert(Replacement.isCanonical()
+         && "replacement types must always be canonical");
+#endif
```
Comment (hahnjo) [inline interpreter/llvm-project/clang/lib/AST/ASTContext.cpp:4604]: As the hack was always enabled, this assertion was almost always disabled - there was one call site that passed `AllowNonCanonical = false` and a couple that used the default parameter, but the most important ones passed `HackForDefaultTemplateArg::AllowNonCanonicalSubst()`. When blindly reverting, I found that this assertion fails from the code path from `areEqualTypes` so I propose to unconditionally disable this.

Now the question is, where do I put this in the monorepo? https://github.com/root-project/llvm-project/commit/660281c8941e3169077d3b9cbde1aeeb1fc66fa5 is currently part of the patches for Cling - shall I add it there? Do we even need this patch for Cling or shall I move this to the patches for ROOT (`ROOT-llvm13`)? @vgvassilev

**Comment by hahnjo** [comment]
> Yep, seems to be all working still:
> 
> ```
> root [0] template <typename T> struct myallocator : public std::allocator<T> {};
> root [1] template <typename T, typename Q = T > class Artificial1 {};
> root [2] template <typename T, typename Q = std::vector<T> > class Artificial2 {};
> root [3] template <typename T, typename Q = std::vector<T, myallocator<T> > > class Artificial3 {};
> root [4] TClass::GetClass("Artificial1<Double32_t>")->GetName()
> (const char *) "Artificial1<Double32_t,Double32_t>"
> root [5] TClass::GetClass("Artificial2<Double32_t>")->GetName()
> (const char *) "Artificial2<Double32_t,vector<Double32_t> >"
> root [6] TClass::GetClass("Artificial3<Double32_t>")->GetName()
> (const char *) "Artificial3<Double32_t,vector<Double32_t,myallocator<Double32_t> > >"
> root [7] TClass::GetClass("Artificial3<Double32_t, std::vector<Double32_t, std::allocator<Double32_t> >>")->GetName()
> (const char *) "Artificial3<Double32_t,vector<Double32_t> >"
> ```

FWIW the output is still the same after reverting the hack in the Clang sources, which was apparently active by default...

File: interpreter/llvm-project/clang/lib/AST/ASTContext.cpp (hunk 2 above)
Comment (vgvassilev) [reply interpreter/llvm-project/clang/lib/AST/ASTContext.cpp:4604]: We do not these patch in cling. We use it mostly to make ROOT I/O optimizations.

File: interpreter/llvm-project/clang/lib/Sema/SemaTemplate.cpp (hunk 1 above)
Comment (vgvassilev) [reply interpreter/llvm-project/clang/lib/Sema/SemaTemplate.cpp:51]: Uh, yeah...

File: interpreter/llvm-project/clang/lib/AST/ASTContext.cpp (hunk 2 above)
Comment (hahnjo) [reply interpreter/llvm-project/clang/lib/AST/ASTContext.cpp:4604]: Ok, I moved the commit in question to the `ROOT-llvm13` branch and will merge once the `llvm-diff` workflow is happy (no need to run the other CI tests again).

**Pull request merged**